### ISSUE TITLE
Anonimization of Kubescape results

### ISSUE THREAD

**Issue by slashben** [kubescape/kubescape#1200]
Feature request by @rawc0der

## Overview
Kubescape report contains fields that can become sensitive information and the user would like to mask/hide them in the reports. In this feature, we would add an option to hide specific fields like namespace/workload names in the report and generate a mapping for later deciphering.

## Problem
User's cluster have PII in the namespace and object names and they need to be treated separately.

**Comment by slashben** [comment]
I suggest discussing it with @rawc0de

**Comment by rakshitgondwal** [comment]
Hey @rawc0der! Where can I reach out to you?

**Comment by rawc0der** [comment]
Hi @rakshitgondwal! Let's maybe move the details to a discussion thread.
I pinged you in there 😎

**Comment by rawc0der** [comment]
Opened Discussions thread https://github.com/kubescape/kubescape/discussions/1201 for brainstorming ideas and design. 

Thanks again  @rakshitgondwal 👋  for jumping on board!

**Comment by nitishchauhan0022** [comment]
@rakshitgondwal are you working on it?

**Comment by rakshitgondwal** [comment]
@nitishchauhan0022 Oh yes, Actually I wanted to discuss this issue with the maintainers once in the community meeting (that would be this Tuesday) on how to continue with this, like the design and stuff.

**Comment by Sanskarzz** [comment]
Hey @rakshitgondwal are you working on this issue?

**Comment by 102vansh** [comment]
can i work on this issue

**Comment by slashben** [comment]
It would be great! Shall I assign it to you?

**Comment by rawc0der** [comment]
I can assist with technical spec and review.

**Comment by 102vansh** [comment]
@slashben yes you can assign

**Comment by rsrajahnsa** [comment]
Is this issue still open ? I would love to work on it.

**Comment by slashben** [comment]
> @slashben  Is this issue still open ? I would love to work on it.

Yes, your welcome to work on it

**Comment by jijo-OO7** [comment]
Hi @slashben @rawc0der 
I came across this issue and it looks quite relevant, especially from a `security/compliance` perspective
Since the issue has been open for a while, I just wanted to check  is this still something the project would like to move forward with?

If yes, I’d be happy to explore a design and take a shot at implementing it.

**Comment by rawc0der** [comment]
Absolutely, go for it! let me know if you need to dry-run any ideas through me before going into implementation mode, I can share more on the original design approach if needed.

**Comment by jijo-OO7** [comment]
> Absolutely, go for it! let me know if you need to dry-run any ideas through me before going into implementation mode, I can share more on the original design approach if needed.

Thanks for confirming!
I’ve been going through the previous discussion https://github.com/kubescape/kubescape/discussions/1201 and will put together a concrete approach. I’ll share it here for feedback before moving into implementation.

**Comment by jijo-OO7** [comment]
hi @rawc0der i have raised a design discussion in #1201  beginning my work on phase-1  will likely push a pr shortly around that ,would be great to have your guidance  as well on the approach

**Comment by jijo-OO7** [comment]
Phase 1 for this feature has now been merged in #2090.

This introduces the initial `--hide` anonymization flow with deterministic pseudonymization for Kubernetes resource identifiers (e.g. names / namespaces / resource references) across scan outputs while preserving report consistency.

I’d like to continue iterating on this issue in follow-up phases to expand coverage toward the broader scope originally discussed here.

Next area I’m planning to explore:

* container image anonymization
* container name anonymization
* other sensitive metadata surfaces that may still leak through scan outputs

If there are any preferences on prioritization or scope for the next phase, happy to align before opening the next PR.

**Comment by matthyx** [comment]
@jijo-OO7 maybe the container image could be sensitive indeed

**Comment by jijo-OO7** [comment]
@matthyx I agree, that aligns with what I was thinking as well.

Container images can definitely expose sensitive internal details (private registries, internal image naming conventions, environment identifiers, etc.), so that feels like a logical next expansion for `--hide`.

I’ll start by exploring the remaining output surfaces that may still leak sensitive identifiers (beginning with container images and container names) and scope the next phase accordingly so it builds cleanly on top of #2090.

**Comment by jijo-OO7** [comment]
Picking this up for further exploration.

The recent container metadata work improved `--hide` coverage, but I’d like to do a broader audit against more production scan scenarios to identify other fields that may still leak identifiable metadata when anonymization is enabled.

The current unit coverage validates the implemented paths, but I want to verify end-to-end behavior using actual scan outputs across realistic workload shapes before proposing additional changes.

I’ll update this thread with findings and follow up with a PR if I identify gaps worth addressing.

**Comment by jijo-OO7** [comment]
A follow-up thought while exploring this further: I think it makes sense to fully close out the **container-related anonymization surface first**, since that’s the area already being expanded incrementally.

The core container identity paths are already covered (`containers`, `initContainers`, `ephemeralContainers` names/images), but there are still a few container-adjacent references that can leak workload / infrastructure identity and probably belong in the default `--hide` behavior as well.

Fields I think should also be anonymized by default to complete the container-related scope:

**Container config references**

* `spec.containers[].env[].valueFrom.secretKeyRef.name`
* `spec.containers[].env[].valueFrom.configMapKeyRef.name`
* `spec.containers[].envFrom[].secretRef.name`
* `spec.containers[].envFrom[].configMapRef.name`

**Container image access identity**

* `spec.imagePullSecrets[].name`

These still reveal internal resource naming / topology even if the container names and images themselves are anonymized.

Then there’s another category that feels more appropriate as **opt-in / user-targeted anonymization**, since hiding them by default may reduce report usefulness:

**User-selective / stricter anonymization**

* literal env values (`spec.containers[].env[].value`)
* container command / args (`spec.containers[].command[]`, `spec.containers[].args[]`)
* custom CRD-specific fields
* eventually user-provided selectors / field targeting

My thought is: finish the container-related identity surface completely first, then expand into broader resource metadata / extensible hide rules in follow-up phases.

Would love feedback if others think any container-related identity paths are missing from the default scope.

**Comment by matthyx** [comment]
@jijo-OO7 makes sense to me, you can hide by default `env`, `envFrom` and `imagePullSecrets`

**Comment by jijo-OO7** [comment]
Hi @matthyx, I’ve raised a follow-up PR for this whenever you have a moment would appreciate your review on #2300.
 Thanks!

**Comment by jijo-OO7** [comment]
While introspecting along the phases  next likely leak surface in `session.ResourceSource`.
`reporthandling.Source` currently carries source metadata such as file paths, relative repo paths, Helm/Kustomize metadata, values provenance, and nested git commit metadata (`hash`, `committer name/email`, commit message). These are currently remapped only by resource ID, but the source payload itself remains untouched, which can still expose internal infrastructure/repository details in hidden scan output.

I’m working on a follow-up PR to extend anonymization coverage for this source metadata while preserving non sensitive fields where appropriate (for example commit date). Will open the PR shortly once the implementation and validation are ready.

If there’s any additional context or specific source metadata you’d like covered as part of this scope, that would be greatly appreciated before I finalize the implementation.

**Comment by jijo-OO7** [comment]
Hi @matthyx 

Quick progress update on the "--hide" anonymization work.

While following up on the earlier 'LocalRootPath' review discussion, I traced the flow through report generation and found that preserving the real path in the session was indeed necessary for downstream consumers (e.g. SARIF generation and FixHandler usage).

To address the remaining leak in hidden output, I've been working on an export-time approach instead of anonymizing the path directly in the session. 

During testing this also uncovered a shallow-copy issue in the report generation path, will be opening a PR shortly once i have something ready .

Once this lands, my plan is to continue with reversible anonymization/mapping support so hidden reports can later be restored to their original values by the report owner. After that, I intend to explore a KMS-backed implementation for secure storage and key management while keeping report formats backward compatible.

Happy to adjust direction if you think a different approach would fit better.

**Comment by matthyx** [comment]
Looks good, thanks @jijo-OO7 !

**Comment by matthyx** [comment]
> Once this lands, my plan is to continue with reversible anonymization/mapping support so hidden reports can later be restored to their original values by the report owner. After that, I intend to explore a KMS-backed implementation for secure storage and key management while keeping report formats backward compatible.

Have you considered borrowing the architectural pattern used by Mozilla SOPS?

Instead of generating separate sidecar mapping files (which can be clunky to track and secure), we could use client-side envelope encryption directly on the exported fields.

How it would look in practice:

1. In-place masking: At export-time, fields like LocalRootPath or sensitive metadata are encrypted using a locally generated Data Encryption Key (DEK). The plain-text value is replaced inline with a cipher string (e.g., LocalRootPath: "ENC[AES256_GCM,...]").
2. Backward Compatibility: Because only the values change, the JSON/YAML schema remains fully intact and valid. Standard consumers can read the unencrypted portions of the report without breaking.
3. KMS-Backed Metadata: Kubescape appends a cryptographic metadata block to the root of the report containing the DEK (encrypted via the user's specified KMS key, like AWS KMS, GCP KMS, or local age key).
4. Seamless Reversibility: When the report owner needs to restore the file locally, they can run a command like kubescape report decrypt report.json, which calls their cloud provider to decrypt the DEK and restore the fields on the fly.

This keeps the output completely self-contained, cryptographically secure, and perfectly backward-compatible.

**Comment by jijo-OO7** [comment]
> Have you considered borrowing the architectural pattern used by Mozilla SOPS?
> 
> Instead of generating separate sidecar mapping files (which can be clunky to track and secure), we could use client-side envelope encryption directly on the exported fields.
> 

Thanks for the suggestion @matthyx. After spending some time thinking through both approaches, I agree that borrowing the SOPS architectural pattern is likely the better longterm direction.

My initial thought was to build reversible anonymization through a mapping layer, but that introduces additional artifacts that need to be stored, secured, transported, versioned, and kept in sync with the report. A self-contained encrypted report avoids those concerns while still providing reversibility.

I also like that the SOPS-style approach naturally preserves backward compatibility since the report schema remains unchanged and only the field values are transformed. It additionally provides a cleaner path toward future KMS-backed key management without requiring a redesign later.

My current plan is to approach this incrementally and start with a small proof of concept that follows the same envelope-encryption model:

* generate a local Data Encryption Key (DEK)
* encrypt selected sensitive report fields in place
* append encryption metadata to the report
* provide a corresponding report decryption flow
* keep the metadata structure extensible for future KMS/age-backed key providers

Once the foundation is validated, support for cloud KMS providers can be added on top without changing the report format.

I'll spend some time putting together a concrete prototype around this direction .

**Comment by matthyx** [comment]
awesome, thanks !

**Comment by jijo-OO7** [comment]
Hi @matthyx,

Following up on the encrypted reports discussion, I've opened a small foundation PR  #2347
that introduces a transformer abstraction for repository metadata along with the initial crypto primitives (AES-GCM + DEK generation).

For now I've intentionally kept the scope limited to `RepoContextMetadata` and a `MappingTransformer` implementation so the existing anonymization behavior remains unchanged while establishing the extension point for future encrypted field handling.

I'd appreciate any feedback on the direction or architecture before moving on to the next step of the encrypted report workflow.

**Comment by jijo-OO7** [comment]
@matthyx  Before I start wiring `EncryptionTransformer` into the repo metadata flow, I wanted to confirm the intended direction.
My current plan is to keep mapping based anonymization as the default path and introduce a separate encrypted repo metadata workflow to validate the end-to-end design first. Key management (KMS) would follow in a separate PR once the encrypted workflow itself is proven

**Comment by jijo-OO7** [comment]
Working on integrating to actual scan worlflow will be updating  once i have something  ready 
Thanks!

**Comment by jijo-OO7** [comment]
> Working on integrating to actual scan worlflow will be updating once i have something ready Thanks!

Hi @matthyx, when you have a moment, could you please take a look at this PR?
PR [ #2425 ] establishes the encrytption and decryption workflow for repometadata  with CLI flags for coresponding functions.

Thanks!

**Comment by jijo-OO7** [comment]
The initial encryption/decryption workflow is now in place for repository metadata.

Looking ahead, the same transformer-based approach can be extended to additional fields currently handled by the anonymization layer. Since these transformations are applied after report generation, encryption coverage can be expanded incrementally while preserving report integrity and existing relationships.

I propose landing the current implementation first and then broadening coverage in follow-up work, focusing on the remaining sensitive metadata and report fields in manageable steps.

**Comment by jijo-OO7** [comment]
Thanks!

I've rebased my current work onto the latest upstream `master`.

PR #2442 carries the follow-up work and now contains only the resource metadata `encryption/decryption` changes.

**Comment by jijo-OO7** [comment]
Hi @matthyx, could you take a look at PR #2493 when you get a moment ,
 it covers support for encrypted resource metadata
Thanks !

**Comment by jijo-OO7** [comment]
Hi @matthyx, now that the encryption/decryption workflow is in place and users provide a 32-byte master key via KUBESCAPE_MASTER_KEY, would you be interested in supporting a passphrase-based workflow as well? We could derive the 32-byte key from a user passphrase (e.g. using a KDF) so users wouldn't need to manage a raw 32-byte key directly. Curious whether you think that's worth adding, or if you'd prefer to keep the current explicit key approach.

**Comment by matthyx** [comment]
I think we can probably leave it there for now. We should probably document the process on kubescape.io and maybe also prepare a blog post?

**Comment by jijo-OO7** [comment]
makes sense Will  start with the documentation, follow up with the kubescape.io docs, and prepare a blog post for the feature.
Thanks for all the reviews and feedback throughout the implementation!

**Comment by jijo-OO7** [comment]
@matthyx I believe all the work for this issue is complete now. Shall we close it?

**Comment by matthyx** [comment]
yes, thanks @jijo-OO7 !
